R/visualize.R

```diff
@@ -0,0 +1,154 @@
+  ggplot(data, aes(x = accuracy, y=metrics, colour=model)) +
+    geom_point() +
+    scale_color_aaas()
+}
+
+fairness_accuracy_tradeoff.ResampleResult <- function(predictions, fairness_measure){
+  predictions = as_benchmark_result(predictions)
+  fairness_accuracy_tradeoff(predcitions, fairness_measure)
+}
+
+
+#' Fairness Comparison
+#'
+#' @param predcitions
+#' TODO: fill it
+#'
+#' @export
+#'
+#' @examples
+#' NULL
+fairness_compare <- function(predcitions, ...){
+  UseMethod("fairness_compare")
+}
+
+fairness_compare.Prediction <- function(predcitions, fairness_measure, data_task){
+  measures = predictions$score(fairness_measure, data_task)
+  data <- melt(data.table(names = names(measures), data = measures), id.vars = "names")
+  ggplot(data, aes(x=names, y=value, fill = names)) +
+    geom_bar(stat = "identity",  width=0.4/length(unique(data$names))) +
+    xlab("metrics") +
+    ylab("values") +
+    theme(legend.position = "none") +
+    scale_fill_hue(c=100, l=60)
+}
+
+fairness_compare.BenchmarkResult <- function(predictions, fairness_measure){
+  fairness_data = predictions$score(fairness_measure)
+  data = melt(data.table(model = fairness_data$learner_id,
+                         fairness_data[,c(11:ncol(fairness_data)), with=FALSE]),
+              id.vars = "model")
+  ggplot(data, aes(x=model, y=value, fill=model)) +
+    geom_bar(stat = "identity",  width=2/length(unique(model))) +
+    xlab("metrics") +
+    ylab("values") +
+    theme(legend.position = "none") +
+    scale_fill_hue(c=100, l=60) +
+    scale_color_aaas() +
+    facet_wrap(~variable)
+}
+
+fairness_compare.ResampleResult <- function(predictions, fairness_measure){
+  predictions = as_benchmark_result(predictions)
+  fairness_compare(predictions, fairness_measure)
+}
+
+#' Fairness Prediction Density Visualization
+#'
+#' @description
+#' This function specializes in visualizing the plots that will demonstrate the predicted probability density for each level in the protected column.
+#' Which will potentially reveal the fairness problems in the dataset if the density differs significantly for each level.
+#' The plot is a combination of boxplot and violin plot. So users could use the plot to detect both the outliers and distribution problems.
+#' The y-axis shows the levels in protected columns. And the x-axis shows the predicted probability.
+#' The title for the plot will demonstrate which class for predicted probability.
+#'
+#' @param predictions (`PredictionClassif()`)\cr
+#' The binary class prediction object that will be evaluated
+#'
+#' @param data_task (`TaskClassif()`)\cr
+#' The data task that contains the protected column.
+#'
+#' @export
+#'
+#' @examples
+#' data_task = tsk("adult_train")
+#' learner = lrn("classif.ranger", predict_type = "prob")
+#' learner$train(data_task)
+#' predictions = learner$predict(data_task)
+#' fairness_prediction_density(predictions, data_task)
+fairness_prediction_density <- function(predictions, data_task){
```

Comment:
This can also be S3 and work across learners.

Reply:
What it gonna look like for benchmark or resampling result? There are multiple predictions, and it gonna look bad if we still plot like PredictionClassif